crates/accelerate/src/synthesis/linear_phase/cz_depth_lnn.rs

```diff
@@ -0,0 +1,172 @@
+
+// A sequence of Lnn gates
+// Represents the return type for Lnn Synthesis algorithms
+pub(crate) type LnnGatesVec = Vec<(StandardGate, SmallVec<[Param; 3]>, SmallVec<[Qubit; 2]>)>;
+
+// A pattern denoted by Pj in [1] for odd number of qubits:
+// [n-2, n-4, n-4, ..., 3, 3, 1, 1, 0, 0, 2, 2, ..., n-3, n-3]
+fn _odd_pattern1(n: usize) -> Vec<usize> {
+    once(n - 2)
+        .chain((0..((n - 3) / 2)).flat_map(|i| [(n - 2 * i - 4); 2]))
+        .chain((0..((n - 1) / 2)).flat_map(|i| [2 * i; 2]))
+        .collect()
+}
+
+// A pattern denoted by Pk in [1] for odd number of qubits:
+// [2, 2, 4, 4, ..., n-1, n-1, n-2, n-2, n-4, n-4, ..., 5, 5, 3, 3, 1]
+fn _odd_pattern2(n: usize) -> Vec<usize> {
+    (0..((n - 1) / 2))
+        .flat_map(|i| [(2 * i + 2); 2])
+        .chain((0..((n - 3) / 2)).flat_map(|i| [n - 2 * i - 2; 2]))
+        .chain(once(1))
+        .collect()
+}
+
+// A pattern denoted by Pj in [1] for even number of qubits:
+// [n-1, n-3, n-3, n-5, n-5, ..., 1, 1, 0, 0, 2, 2, ..., n-4, n-4, n-2]
+fn _even_pattern1(n: usize) -> Vec<usize> {
+    once(n - 1)
+        .chain((0..((n - 2) / 2)).flat_map(|i| [n - 2 * i - 3; 2]))
+        .chain((0..((n - 2) / 2)).flat_map(|i| [2 * i; 2]))
+        .chain(once(n - 2))
+        .collect()
+}
+
+// A pattern denoted by Pk in [1] for even number of qubits:
+// [2, 2, 4, 4, ..., n-2, n-2, n-1, n-1, ..., 3, 3, 1, 1]
+fn _even_pattern2(n: usize) -> Vec<usize> {
+    (0..((n - 2) / 2))
+        .flat_map(|i| [2 * (i + 1); 2])
+        .chain((0..(n / 2)).flat_map(|i| [(n - 2 * i - 1); 2]))
+        .collect()
+}
+
+// Creating the patterns for the phase layers.
+fn _create_patterns(n: usize) -> HashMap<(usize, usize), (usize, usize)> {
+    let (pat1, pat2) = if n % 2 == 0 {
+        (_even_pattern1(n), _even_pattern2(n))
+    } else {
+        (_odd_pattern1(n), _odd_pattern2(n))
+    };
+
+    let ind = if n % 2 == 0 {
+        (2 * n - 4) / 2
+    } else {
+        (2 * n - 4) / 2 - 1
+    };
+
+    HashMap::from_iter((0..n).map(|i| ((0, i), (i, i))).chain(
+        (0..(n / 2)).cartesian_product(0..n).map(|(layer, i)| {
+            (
+                (layer + 1, i),
+                (pat1[ind - (2 * layer) + i], pat2[(2 * layer) + i]),
+            )
+        }),
+    ))
+}
+
+// Appends correct phase gate during CZ synthesis
+fn _append_phase_gate(pat_val: usize, gates: &mut LnnGatesVec, qubit: usize) {
+    // Add phase gates: s, sdg or z
+    let gate_id = pat_val % 4;
+    if gate_id != 0 {
+        let gate = match gate_id {
+            1 => StandardGate::SdgGate,
+            2 => StandardGate::ZGate,
+            3 => StandardGate::SGate,
+            // cover 2 and 3
+            _ => unreachable!(), // unreachable as we have modulo 4
+        };
+        gates.push((gate, smallvec![], smallvec![Qubit(qubit as u32)]));
+    }
+}
+
+// Synthesis of a CZ circuit for linear nearest neighbor (LNN) connectivity,
+// based on Maslov and Roetteler.
```

Comment:
A general comment. Usually we use triple slashes (i.e. `///`) for function docstrings and double slashes (i.e. `//`) for regular comments.  I believe that it makes sense to promote this comment here to a docstring. The same comment applies in many other places.